Dashr financial report

HammadTheOne opened the pull request:
Issue for app: #[issue number here]

# App pull request
- [ ] This is a new app
- [x] I am improving an existing app (redesigns/code "makeovers")

## About

- Playground deployment URL (new version): https://dash-playground.plotly.host/Manager/apps/dashr-financial-report/overview
- Current gallery app URL: https://dash-gallery.plotly.host/dash-vanguard-report/
- Python app repository link: https://github.com/plotly/dash-vanguard-report

## Workflow

- [x] I have created a branch in the appropriate monorepo, and the
  elements necessary for successful deployment are in place.
- [x] If the app is a redesigned and/or restyled version of an
  existing gallery app, I've summarized the changes requested in the
  appropriate Streambed issue and confirm that they have been applied.
- [ ] If the app is on the Dash Gallery portal, I have added a link to
  the GitHub repository for the source code in the portal description.
- [x] If the app is a reimplementation of a Python gallery app for the
  DashR gallery, the app in this PR mimics, as closely as possible,
  the style and functionality of the existing app.

## The pre-review review

I have addressed all of the following questions:

- [x] Does everything in my code serve some purpose? (I have removed
  any dead and/or irrelevant code.)
- [x] Does everything in my code have a clear purpose? (My code is
  readable and, where it isn't, it has been commented appropriately.)]
- [x] Am I reinventing the wheel? (I have used appropriate packages to
  lessen the volume of code that needs to be maintained.)

Scott-Clone commented:
Code Review:

- move external CSS links to assets folder 
- remove extra spacing throughout code 
- Dash Financial logo does not display properly 

![Screen Shot 2019-06-19 at 1 50 07 PM](https://user-images.githubusercontent.com/25450911/59795165-a57b6800-92a8-11e9-9207-fca5637f544e.png)

- Mobile view does not work properly in chrome mobile view, but works when window is manually resized
- Fix spacing for "Hypothetical Growth" plot, margin not aligned with other plots, check margin/padding spacing throughout code
- Check code style throughout code (check with lintr)
- General presentation: In "Minimums" table, Initial dollar amount is presented in quotations in both versions of the app, could be worth converting to numeric?

![Screen Shot 2019-06-19 at 2 50 18 PM](https://user-images.githubusercontent.com/25450911/59795171-a8765880-92a8-11e9-88e8-546e803e0e05.png)

- Fix gains table

![Screen Shot 2019-06-19 at 3 59 43 PM](https://user-images.githubusercontent.com/25450911/59796601-e0cb6600-92ab-11e9-9eda-ccd488b4d44b.png)


@rpkyle

HammadTheOne commented:
> Code Review:
> 
>     * move external CSS links to assets folder
> 
>     * remove extra spacing throughout code
> 
>     * Dash Financial logo does not display properly
> 
> 
> ![Screen Shot 2019-06-19 at 1 50 07 PM](https://user-images.githubusercontent.com/25450911/59795165-a57b6800-92a8-11e9-9207-fca5637f544e.png)
> 
>     * Mobile view does not work properly in chrome mobile view, but works when window is manually resized
> 
>     * Fix spacing for "Hypothetical Growth" plot, margin not aligned with other plots, check margin/padding spacing throughout code
> 
>     * Check code style throughout code (check with lintr)
> 
>     * General presentation: In "Minimums" table, Initial dollar amount is presented in quotations in both versions of the app, could be worth converting to numeric?
> 
> 
> ![Screen Shot 2019-06-19 at 2 50 18 PM](https://user-images.githubusercontent.com/25450911/59795171-a8765880-92a8-11e9-88e8-546e803e0e05.png)
> 
>     * Fix gains table
> 
> 
> ![Screen Shot 2019-06-19 at 3 59 43 PM](https://user-images.githubusercontent.com/25450911/59796601-e0cb6600-92ab-11e9-9eda-ccd488b4d44b.png)

test
> 


test

HammadTheOne commented:
*Fixed Dash Financial Logo dimensions. 

*Mobile view should be fixed on an actual mobile device, needs testing once deployed to Playground.

*"Minimum's" table fixed, now shows only dollar values, no string quotes. 

*"Gain's Table" issue is an encoding issue, certain character's show up as that. Need to check on Playground deployment, may be resolved.

*"Hypothetical Growth" graph is now properly aligned with margins.

*External CSS removed and added to `assets` folder.

rpkyle commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,130 @@
+# Install a package that is necessary and has not been installed, or load one that is.
+
+usePackage <- function(p) 
```
@HammadTheOne Where did this function come from?

HammadTheOne replied:
🔪, remnant of the installation process before knowing about `init.R` file.

rpkyle commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,130 @@
+  
+  rows <- lapply(seq(1, n), function(i) {
+    
+    htmlTr(children = lapply(as.character(df[i,]), htmlTd))
```
👍

rpkyle commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,130 @@
+
+# Remove.Factors
+
+remove.factors <- function (df) 
```
Not 100% certain, but I suspect you could simplify this, e.g.

```
f2n <- function(x) as.numeric(levels(x))[x]
f2c <- function(x) as.character(levels(x))[x]
```

HammadTheOne replied:
Simplification worked, no notable changes with how it affects the tables, but the code is a lot cleaner.

rpkyle commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,130 @@
+      htmlDiv(list(
+        grey_line
+      )),
+      htmlBr(list()),
```
Hmm. Does this require an empty list be passed? What happens when it's invoked with no arguments?

HammadTheOne replied:
It does not. Fixed.

rpkyle commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,130 @@
+                        )))),
+    
+    htmlDiv(list(htmlA(
+      htmlImg(src='https://user-images.githubusercontent.com/1865834/50180824-abcc5f80-02d8-11e9-8319-8842909c3f8e.png', 
```
I'd use `assets` for this, store the image locally.

HammadTheOne replied:
Fixed, updated with new Plotly logo.

sacul-git commented on apps/dashr-financial-report/app.R:
```diff
@@ -0,0 +1,1093 @@
+
+#################################################################################
+
+colors <- list(
```
It doesn't matter for your app, but using `colors` as a name masks `grDevices::colors`, best to avoid when possible (though honestly it seems like most of the R community doesn't worry about overwriting variables from package to package)

HammadTheOne replied:
Removed it regardless, turns out it was unnecessary lines anyways.

sacul-git commented on apps/dashr-financial-report/app.R:
```diff
@@ -0,0 +1,1093 @@
+
+performance_data$years <- factor(performance_data$years, levels = performance_data$years[order(c(1,2,3,4,5))])
+
+performance <- plot_ly(performance_data, x= ~years, y= ~Index_Fund, type = "bar",
```
Long lines like this might be more digestible if you format them differently, maybe along the lines of:

```
performance <- plot_ly(
  performance_data, x= ~years, y= ~Index_Fund, 
  type = "bar", name = 'Calibre Index Fund', 
  width = 340,  height = 200) %>%
    add_trace(y= ~SP_Index, name = "S&P Index Fund") %>%
    layout(
      yaxis = ax, xaxis = ax, 
      colorway = c('#98151B', '#DCDCDC'), 
      legend = list(
        x = 1.0, y =  0.95, 
        orientation = 'h', 
        yanchor = "top", 
        font = list(size =9)
      ),
      autosize = FALSE, 
      bargap = 0.35, 
      hovermode = "closest",  
      margin = list(r=0, t=20, b=10, l=10)
    )
```

HammadTheOne replied:
Fixed, adjusting formatting of all lines exceeding that length (except for paragraph text for `dccMarkdown` formatting purposes).

sacul-git commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,122 @@
+# Generate a Table Function 
+generate_table <- function(df, nrows=20)
+  
+  # function generates a dash table from a supplied data frame (df)
```
Should maybe update this comment to replace `dash table` with just `table` to avoid confusion, since this is giving an `htmlTable` rather than a `dashTable`

HammadTheOne replied:
Fixed, replaced with htmlTable.

sacul-git commented on apps/dashr-financial-report/assets/FinancialFunctions.R:
```diff
@@ -0,0 +1,122 @@
+                        htmlButton(list("Full Version"), className = "redbutton"
+                        )))),
+    
+    htmlDiv(list(htmlA( href = 'https://github.com/plotly/dash-financial-report',
```
This should probably point to the monorepo instead of the old standalone repo

HammadTheOne replied:
Fixed.
